### DISCUSSION TITLE
How metrics are computed?

### DISCUSSION THREAD

**Discussion by William9Baker** [open-edge-platform/anomalib#1203] [Q&A]
When it comes to 'trainer.fit(model=model, datamodule=datamodule)', how does it automatically jump to this position that 'fpr, tpr = self._compute()' in the file of 'aupro.py'?

**Comment by samet-akcay** [comment]
@William9Baker, can you elaborate your question?

**Comment by samet-akcay** [comment]
`model` that we pass to `Trainer` is always a subclass of `AnomalyModule` that contains all these metric and threshold values as attributes, which are shown below:
https://github.com/openvinotoolkit/anomalib/blob/93239858c51de63dba064bfc6b1e66522d2586ae/src/anomalib/models/components/base/anomaly_module.py#L32-L54

When the validation is complete, Lightning's `Trainer` object calls `_collect_outputs` methods which computes and updates the metrics.
https://github.com/openvinotoolkit/anomalib/blob/93239858c51de63dba064bfc6b1e66522d2586ae/src/anomalib/models/components/base/anomaly_module.py#L138-L147

`self._collect_outputs` method does the following to compute the metrics:
https://github.com/openvinotoolkit/anomalib/blob/93239858c51de63dba064bfc6b1e66522d2586ae/src/anomalib/models/components/base/anomaly_module.py#L172-L182
